PSOC6: enable export to CMake

vmedcy opened the pull request:
### Description

CMake exporter needs to select a single hex file with CM0+ prebuilt image from resources.
The approach for the hex_files subset selection is identical to makefile exporter: https://github.com/ARMmbed/mbed-os/pull/9466

Single hex file should be passed to srec_cat when "hex_filename" is set in targets.json or mbed_app.json.

There is no impact on non-PSoC6 targets (without "hex_filename" attribute in targets.json).

### Pull request type

    [ ] Fix
    [X] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@theotherjimmy 

### Release Notes

vmedcy committed:
PSOC6: enable export to CMake

The approach for the hex_files subset selection is identical
to makefile exporter: https://github.com/ARMmbed/mbed-os/pull/9466

Single hex file should be passed to srec_cat when hex_filename
is set in targets.json or mbed_app.json.

theotherjimmy commented on tools/export/cmake/__init__.py:
```diff
@@ -74,6 +75,12 @@ def generate(self):
+        hex_files = self.resources.hex_files
+        if hasattr(self.toolchain.target, 'hex_filename'):
+            hex_filename = self.toolchain.target.hex_filename
+            hex_files = list(f for f in hex_files if basename(f) == hex_filename)
```
This is the same thing:

```suggestion
            hex_files = [f for f in hex_files if basename(f) == hex_filename]
```

vmedcy replied:
Thanks for suggestion, will use the syntax next time I touch the exporter code.
For now I'd like to keep the code in cmake and makefile exporters in sync.

theotherjimmy replied:
Should it be a function on the base exporter class then? That way there's no need to keep them in sync.

vmedcy replied:
I can add new property to the base Exporter class, similar to existing [libraries ](https://github.com/ARMmbed/mbed-os/blob/mbed-os-5.11.4/tools/export/exporters.py#L122)property, and update cmake and makefile exporters to use it. Please confirm this is correct:
```python
@property
def hex_files(self):
    hex_files = self.resources.hex_files
    if hasattr(self.toolchain.target, 'hex_filename'):
        hex_filename = self.toolchain.target.hex_filename
        hex_files = [f for f in hex_files if basename(f) == hex_filename]
    return hex_files
```

If so, I will update this PR with the proposed change and update both exporters to use it.

vmedcy replied:
I pushed https://github.com/ARMmbed/mbed-os/pull/9756/commits/cf0ea921046c4c4eb8102f7e9cc72dec86ec6242 and tested it has no functional impact (verified `mbed export -m CY8CKIT_062_BLE -i cmake_gcc_arm` and `mbed export -m CY8CKIT_062_BLE -i eclipse_gcc_arm`).

I also used [] to initalize the python list.

Please review. I updated the PR description to mention this is [X] Refactor

vmedcy committed:
tools/export: move hex_files selection to base Exporter class

CMake and makefile exporters share a common logic for hex file
selection. Factor it as a common property in the base class
to avoid code duplication.
